src/OrchardCore/OrchardCore.Infrastructure/Shell/ShellDescriptorManager.cs

```diff
@@ -98,32 +93,26 @@ public async Task UpdateShellDescriptorAsync(int priorSerialNumber, IEnumerable<
                 _logger.LogInformation("Updating shell descriptor for tenant '{TenantName}' ...", _shellSettings.Name);
             }
 
-            if (shellDescriptorRecord == null)
-            {
-                shellDescriptorRecord = new ShellDescriptor { SerialNumber = 1 };
-            }
-            else
-            {
-                shellDescriptorRecord.SerialNumber++;
-            }
-
-            shellDescriptorRecord.Features = _alwaysEnabledFeatures.Union(enabledFeatures).ToList();
-            shellDescriptorRecord.Installed = shellDescriptorRecord.Installed.Union(shellDescriptorRecord.Features).ToList();
+            shellDescriptor.SerialNumber++;
+            shellDescriptor.Features = _alwaysEnabledFeatures.Union(enabledFeatures).ToList();
+            shellDescriptor.Installed = shellDescriptor.Installed.Union(shellDescriptor.Features).ToList();
 
             if (_logger.IsEnabled(LogLevel.Information))
             {
                 _logger.LogInformation("Shell descriptor updated for tenant '{TenantName}'.", _shellSettings.Name);
             }
 
-            _session.Save(shellDescriptorRecord);
+            await _documentStore.UpdateAsync(shellDescriptor, (_) => Task.CompletedTask);
 
             // In the 'ChangedAsync()' event the shell will be released so that, on request, a new one will be built.
             // So, we commit the session earlier to prevent a new shell from being built from an outdated descriptor.
-
-            await _session.SaveChangesAsync();
-
             await _shellDescriptorManagerEventHandlers.InvokeAsync((handler, shellDescriptorRecord, _shellSettings) =>
-                handler.ChangedAsync(shellDescriptorRecord, _shellSettings), shellDescriptorRecord, _shellSettings, _logger);
+                handler.ChangedAsync(shellDescriptorRecord, _shellSettings), shellDescriptor, _shellSettings, _logger);
+        }
+
+        private static Task<ShellDescriptor> ShellDescriptorFactory()
+        {
+            return Task.FromResult(new ShellDescriptor { SerialNumber = 0 });
```

Comment:
Define it as static field

```csharp
private static readonly ShellDescriptor _shellDescriptor = new ShellDescriptor { SerialNumber = 0 };
```

Reply:
Then react to the rename change

Reply:
Could you explain me why ?
If `_shellDescriptor` is readonly, we can't replace it with the value retrieved from store ?
And if it's static, it means we have one instance for all tenants ?

Reply:
I see, just forgot my comments